yourbench/pipeline/multi_hop_question_generation.py

```diff
@@ -0,0 +1,510 @@
+2. Prompts a large language model to generate multi-hop question-answer pairs.
+3. Parses and saves the generated questions in a structured dataset.
+
+Error Handling and Logging:
+---------------------------
+- Comprehensive logging is performed using `loguru` at various levels to trace execution.
+- Exceptions are caught and logged as errors, with the module attempting to continue
+  where practical.
+- Critical issues produce warnings or errors and gracefully terminate the stage.
+
+Module-Level Dependencies:
+--------------------------
+- Relies on the shared pipeline utilities (e.g., `yourbench.utils.dataset_engine`,
+  `yourbench.utils.inference_engine`, `yourbench.utils.prompts`).
+- Preserves the existing signature and functionality for downstream consistency.
+"""
+
+import json
+import random
+import re
+from dataclasses import dataclass, field
+from typing import Any, Dict, List
+
+from datasets import Dataset
+from loguru import logger
+
+from yourbench.utils.dataset_engine import (
+    save_dataset,
+    smart_get_output_dataset_name,
+    smart_get_output_subset,
+    smart_get_source_dataset_name,
+    smart_get_source_subset,
+    smart_load_dataset,
+)
+from yourbench.utils.inference_engine import InferenceCall, run_inference
+from yourbench.utils.prompts import (
+    MULTI_HOP_QUESTION_GENERATION_SYSTEM_PROMPT,
+    MULTI_HOP_QUESTION_GENERATION_USER_PROMPT,
+)
+
+
+@dataclass
+class MultiHopQuestionRow:
+    """
+    Data structure to represent a single multi-hop question row.
+
+    Attributes:
+        document_id (str):
+            Identifier of the document from which the question is generated.
+        source_chunk_ids (List[str]):
+            List of single-hop chunk IDs used in producing the multi-hop question.
+        question (str):
+            The generated multi-hop question text.
+        self_answer (str):
+            A plausible model-provided answer to the generated question.
+        estimated_difficulty (int):
+            Difficulty rating on a scale of 1 (easiest) to 10 (hardest).
+        self_assessed_question_type (str):
+            A descriptor for the question style (e.g., "analytical", "factual").
+        generating_model (str):
+            Name of the model that generated the question-answer pair.
+        thought_process (str):
+            Free-form text describing the model's reasoning for the question.
+        citations (List[str]):
+            Optional references/quotations from the combined chunks.
+        raw_response (str):
+            The full, unedited response from the model.
+    """
+
+    document_id: str
+    source_chunk_ids: List[str]
+    question: str
+    self_answer: str
+    estimated_difficulty: int
+    self_assessed_question_type: str
+    generating_model: str
+    thought_process: str
+    citations: List[str] = field(default_factory=list)
+    raw_response: str = field(default="")
+
+
+def run(config: Dict[str, Any]) -> None:
+    """
```

Comment:
this fn is way too long - it would make sense to split it in logical blocks, like
multihop_dataset_loading
multihop_chunk_sampling
multihop_qa_generation
parsing
saving